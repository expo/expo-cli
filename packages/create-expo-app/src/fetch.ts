import fetch from 'node-fetch';
import os from 'os';
import path from 'path';

import { FileSystemCache } from './cache/FileSystemCache';
import { wrapFetchWithCache } from './cache/wrapFetchWithCache';

const debug = require('debug')('create-expo-app:fetch') as typeof console.log;

export function getCacheFilePath(subdir: string = 'template-cache') {
  // TODO: Revisit
  return path.join(os.homedir(), '.create-expo-app', subdir);
}

export function createFetch({
  cacheDirectory,
  ttl,
}: { cacheDirectory?: string; ttl?: number | false } = {}) {
  if (ttl === false) {
    return fetch;
  }
  const directory = getCacheFilePath(cacheDirectory);
  debug('Creating fetch with cache directory:', directory);
  return wrapFetchWithCache(
    fetch,
    new FileSystemCache({
      cacheDirectory: directory,
      ttl,
    })
  );
}
