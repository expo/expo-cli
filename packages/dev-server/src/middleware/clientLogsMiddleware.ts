import http from 'http';
import Log from '@expo/bunyan';

type ConsoleLogLevel = 'info' | 'warn' | 'error' | 'debug';

export default function clientLogsMiddleware(logger: Log) {
  return function(
    req: http.IncomingMessage & { body?: any },
    res: http.ServerResponse,
    next: (err?: any) => void
  ) {
    try {
      const deviceId = req.headers['device-id'];
      const deviceName = req.headers['device-name'];
      if (!deviceId) {
        res.writeHead(400).end('Missing Device-Id.');
        return;
      }
      if (!deviceName) {
        res.writeHead(400).end('Missing Device-Name.');
        return;
      }
      if (!req.body) {
        res.writeHead(400).end('Missing request body.');
        return;
      }
      handleDeviceLogs(logger, deviceId.toString(), deviceName.toString(), req.body);
    } catch (error) {
      logger.error({ tag: 'expo' }, `Error getting device logs: ${error} ${error.stack}`);
      next(error);
    }
    res.end('Success');
  };
}

function isIgnorableBugReportingExtraData(body: any[]) {
  return body.length === 2 && body[0] === 'BugReporting extraData:';
}

function isAppRegistryStartupMessage(body: any[]) {
  return (
    body.length === 1 &&
    (/^Running application "main" with appParams:/.test(body[0]) ||
      /^Running "main" with \{/.test(body[0]))
  );
}

function handleDeviceLogs(logger: Log, deviceId: string, deviceName: string, logs: any) {
  for (let i = 0; i < logs.length; i++) {
    const log = logs[i];
    let body = typeof log.body === 'string' ? [log.body] : log.body;
    let { level } = log;

    if (isIgnorableBugReportingExtraData(body)) {
      level = 'debug';
    }
    if (isAppRegistryStartupMessage(body)) {
      body = [`Running application on ${deviceName}.`];
    }

    const args = body.map((obj: any) => {
      if (typeof obj === 'undefined') {
        return 'undefined';
      }
      if (obj === 'null') {
        return 'null';
      }
      if (typeof obj === 'string' || typeof obj === 'number' || typeof obj === 'boolean') {
        return obj;
      }
      try {
        return JSON.stringify(obj);
      } catch (e) {
        return obj.toString();
      }
    });
    const logLevel =
      level === 'info' || level === 'warn' || level === 'error' || level === 'debug'
        ? (level as ConsoleLogLevel)
        : 'info';
    logger[logLevel](
      {
        tag: 'device',
        deviceId,
        deviceName,
        groupDepth: log.groupDepth,
        shouldHide: log.shouldHide,
        includesStack: log.includesStack,
      },
      ...args
    );
  }
}
