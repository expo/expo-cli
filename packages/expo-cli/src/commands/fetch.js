/**
 * @flow
 */

import chalk from 'chalk';
import fs from 'fs';
import path from 'path';
import { Credentials, Exp } from 'xdl';
import crypto from 'crypto';
import spawnAsync from '@expo/spawn-async';
import log from '../log';

function exportCertAsync(keystoreFile, keystorePassword, keyAlias, certFile) {
  return spawnAsync('keytool', [
    '-exportcert',
    '-keystore',
    keystoreFile,
    '-storepass',
    keystorePassword,
    '-alias',
    keyAlias,
    '-file',
    certFile,
    '-noprompt',
    '-storetype',
    'JKS',
  ]);
}

export default (program: any) => {
  program
    .command('fetch:ios:certs [project-dir]')
    .description(
      `Fetch this project's iOS certificates and provisioning profile. Writes certificates to PROJECT_DIR/PROJECT_NAME_(dist|push).p12 and prints passwords to stdout.`
    )
    .asyncActionProjectDir(async (projectDir, options) => {
      const {
        args: {
          username,
          remotePackageName,
          remoteFullPackageName: experienceName,
          bundleIdentifierIOS: bundleIdentifier,
        },
      } = await Exp.getPublishInfoAsync(projectDir);

      let distOutputFile = path.resolve(projectDir, `${remotePackageName}_dist.p12`);
      let pushOutputFile = path.resolve(projectDir, `${remotePackageName}_push.p12`);

      const credentialMetadata = { username, experienceName, platform: 'ios', bundleIdentifier };

      log(`Retreiving iOS credentials for ${experienceName}`);

      try {
        const {
          certP12,
          certPassword,
          certPrivateSigningKey,
          pushP12,
          pushPassword,
          pushPrivateSigningKey,
          provisioningProfile,
          teamId,
        } = await Credentials.getCredentialsForPlatform(credentialMetadata);
        // if undefines because some people might have pre-local-auth as default credentials.
        if (teamId !== undefined) {
          log(`These credentials are associated with Apple Team ID: ${teamId}`);
        }
        log(`Writing distribution cert to ${distOutputFile}...`);
        fs.writeFileSync(distOutputFile, Buffer.from(certP12, 'base64'));
        if (certPrivateSigningKey !== undefined) {
          let keyPath = path.resolve(projectDir, `${remotePackageName}_dist_cert_private.key`);
          fs.writeFileSync(keyPath, certPrivateSigningKey);
        }
        log('Done writing distribution cert credentials to disk.');
        log(`Writing push cert to ${pushOutputFile}...`);
        fs.writeFileSync(pushOutputFile, Buffer.from(pushP12, 'base64'));
        if (pushPrivateSigningKey !== undefined) {
          let keyPath = path.resolve(projectDir, `${remotePackageName}_push_cert_private.key`);
          fs.writeFileSync(keyPath, pushPrivateSigningKey);
        }
        log('Done writing push cert credentials to disk.');
        if (provisioningProfile !== undefined) {
          let p = path.resolve(projectDir, `${remotePackageName}.mobileprovision`);
          log(`Writing provisioning profile to ${p}...`);
          fs.writeFileSync(p, Buffer.from(provisioningProfile, 'base64'));
          log('Done writing provisioning profile to disk');
        }
        log(`Save these important values as well:

Distribution p12 password: ${chalk.bold(certPassword)}
Push p12 password:         ${chalk.bold(pushPassword)}
`);
      } catch (e) {
        throw new Error('Unable to fetch credentials for this project. Are you sure they exist?');
      }

      log('All done!');
    }, true);

  program
    .command('fetch:android:keystore [project-dir]')
    .description(
      "Fetch this project's Android keystore. Writes keystore to PROJECT_DIR/PROJECT_NAME.jks and prints passwords to stdout."
    )
    .asyncActionProjectDir(async (projectDir, options) => {
      const {
        args: { username, remotePackageName, remoteFullPackageName: experienceName },
      } = await Exp.getPublishInfoAsync(projectDir);

      const backupKeystoreOutputPath = path.resolve(projectDir, `${remotePackageName}.jks`);
      await Credentials.backupExistingAndroidCredentials({
        outputPath: backupKeystoreOutputPath,
        username,
        experienceName,
        log,
      });
    }, true);

  program
    .command('fetch:android:hashes [project-dir]')
    .description(
      "Fetch this project's Android key hashes needed to setup Google/Facebook authentication."
    )
    .asyncActionProjectDir(async (projectDir, options) => {
      const {
        args: { username, remotePackageName, remoteFullPackageName: experienceName },
      } = await Exp.getPublishInfoAsync(projectDir);

      const outputPath = path.resolve(projectDir, `${remotePackageName}.tmp.jks`);
      const { keystorePassword, keyAlias } = await Credentials.backupExistingAndroidCredentials({
        outputPath,
        username,
        experienceName,
        log,
        logSecrets: false,
      });
      const certFile = outputPath.replace('jks', 'cer');
      try {
        await exportCertAsync(outputPath, keystorePassword, keyAlias, certFile);
        const data = fs.readFileSync(certFile);
        const googleHash = crypto
          .createHash('sha1')
          .update(data)
          .digest('hex')
          .toUpperCase();
        const googleHash256 = crypto
          .createHash('sha256')
          .update(data)
          .digest('hex')
          .toUpperCase();
        const fbHash = crypto
          .createHash('sha1')
          .update(data)
          .digest('base64');
        log(`Google Certificate Fingerprint:     ${googleHash.replace(/(.{2}(?!$))/g, '$1:')}`);
        log(`Google Certificate Hash (SHA-1):    ${googleHash}`);
        log(`Google Certificate Hash (SHA-256):  ${googleHash256}`);
        log(`Facebook Key Hash:                  ${fbHash}`);
      } catch (err) {
        if (err.code === 'ENOENT') {
          log.warn('Are you sure you have keytool installed?');
          log('keytool is part of openJDK: http://openjdk.java.net/');
          log('Also make sure that keytool is in your PATH after installation.');
        }
        if (err.stdout) {
          log(err.stdout);
        }
        if (err.stderr) {
          log.error(err.stderr);
        }
        throw err;
      } finally {
        try {
          fs.unlinkSync(certFile);
        } catch (err) {
          if (err.code !== 'ENOENT') {
            log.error(err);
          }
        }
        try {
          fs.unlinkSync(outputPath);
        } catch (err) {
          if (err.code !== 'ENOENT') {
            log.error(err);
          }
        }
      }

      log('All done!');
    }, true);
};
