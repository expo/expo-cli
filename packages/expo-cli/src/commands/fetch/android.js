import path from 'path';
import fs from 'fs';

import { Credentials, Exp } from '@expo/xdl';

import log from '../../log';

async function fetchAndroidKeystoreAsync(projectDir) {
  const {
    args: { username, remotePackageName, remoteFullPackageName: experienceName },
  } = await Exp.getPublishInfoAsync(projectDir);

  const backupKeystoreOutputPath = path.resolve(projectDir, `${remotePackageName}.jks`);
  await Credentials.Android.backupExistingCredentials(
    {
      outputPath: backupKeystoreOutputPath,
      username,
      experienceName,
    },
    log
  );
}

async function fetchAndroidHashesAsync(projectDir) {
  const {
    args: { username, remotePackageName, remoteFullPackageName: experienceName },
  } = await Exp.getPublishInfoAsync(projectDir);

  const outputPath = path.resolve(projectDir, `${remotePackageName}.tmp.jks`);
  try {
    const { keystorePassword, keyAlias } = await Credentials.Android.backupExistingCredentials(
      {
        outputPath,
        username,
        experienceName,
      },
      log,
      false
    );
    await Credentials.Android.logKeystoreHashes(
      {
        keystorePath: outputPath,
        keystorePassword,
        keyAlias,
      },
      log
    );
  } finally {
    try {
      fs.unlinkSync(outputPath);
    } catch (err) {
      if (err.code !== 'ENOENT') {
        log.error(err);
      }
    }
  }
}

async function fetchAndroidUploadCertAsync(projectDir) {
  const {
    args: { username, remotePackageName, remoteFullPackageName: experienceName },
  } = await Exp.getPublishInfoAsync(projectDir);

  const keystorePath = path.resolve(projectDir, `${remotePackageName}.tmp.jks`);
  const uploadKeyPath = path.resolve(projectDir, `${remotePackageName}_upload_cert.pem`);
  try {
    const { keystorePassword, keyAlias } = await Credentials.Android.backupExistingCredentials(
      {
        outputPath: keystorePath,
        username,
        experienceName,
      },
      log,
      false
    );

    log(`Writing upload key to ${uploadKeyPath}`);
    await Credentials.Android.exportCertBase64(
      keystorePath,
      keystorePassword,
      keyAlias,
      uploadKeyPath
    );
  } finally {
    try {
      fs.unlinkSync(keystorePath);
    } catch (err) {
      if (err.code !== 'ENOENT') {
        log.error(err);
      }
    }
  }
}

export { fetchAndroidKeystoreAsync, fetchAndroidHashesAsync, fetchAndroidUploadCertAsync };
