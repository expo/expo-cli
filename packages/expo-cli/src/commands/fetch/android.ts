import path from 'path';
import fs from 'fs-extra';
import get from 'lodash/get';

import { AndroidCredentials } from '@expo/xdl';
import { DownloadKeystore } from '../../credentials/views/AndroidCredentials';
import { Context } from '../../credentials';

import log from '../../log';

async function maybeRenameExistingFile(projectDir: string, filename: string) {
  let desiredFilePath = path.resolve(projectDir, filename);

  if (await fs.pathExists(desiredFilePath)) {
    let num = 1;
    while (await fs.pathExists(path.resolve(projectDir, `OLD_${num}_${filename}`))) {
      num++;
    }
    log(
      `\nA file already exists at "${desiredFilePath}"\n  Renaming the existing file to OLD_${num}_${filename}\n`
    );
    await fs.rename(desiredFilePath, path.resolve(projectDir, `OLD_${num}_${filename}`));
  }
}

export async function fetchAndroidKeystoreAsync(projectDir: string): Promise<void> {
  const ctx = new Context();
  await ctx.init(projectDir);

  const keystoreFilename = `${ctx.manifest.slug}.jks`;
  await maybeRenameExistingFile(projectDir, keystoreFilename);
  const backupKeystoreOutputPath = path.resolve(projectDir, keystoreFilename);

  const view = new DownloadKeystore(ctx.manifest.slug);
  await view.fetch(ctx);
  await view.save(ctx, backupKeystoreOutputPath, true);
}

export async function fetchAndroidHashesAsync(projectDir: string): Promise<void> {
  const ctx = new Context();
  await ctx.init(projectDir);
  const outputPath = path.resolve(projectDir, `${ctx.manifest.slug}.tmp.jks`);
  try {
    const view = new DownloadKeystore(ctx.manifest.slug);
    await view.fetch(ctx);
    await view.save(ctx, outputPath);

    // @ts-ignore: keyPassword isn't defined
    await AndroidCredentials.logKeystoreHashes({
      keystorePath: outputPath,
      keystorePassword: get(view, 'credentials.keystorePassword'),
      keyAlias: get(view, 'credentials.keyAlias'),
    });
    log(
      `\nNote: if you are using Google Play signing, this app will be signed with a different key after publishing to the store, and you'll need to use the hashes displayed in the Google Play console.`
    );
  } finally {
    try {
      fs.unlink(outputPath);
    } catch (err) {
      if (err.code !== 'ENOENT') {
        log.error(err);
      }
    }
  }
}

export async function fetchAndroidUploadCertAsync(projectDir: string): Promise<void> {
  const ctx = new Context();
  await ctx.init(projectDir);

  const keystorePath = path.resolve(projectDir, `${ctx.manifest.slug}.tmp.jks`);

  const uploadKeyFilename = `${ctx.manifest.slug}_upload_cert.pem`;
  await maybeRenameExistingFile(projectDir, uploadKeyFilename);
  const uploadKeyPath = path.resolve(projectDir, uploadKeyFilename);

  try {
    const view = new DownloadKeystore(ctx.manifest.slug);
    await view.fetch(ctx);
    await view.save(ctx, keystorePath);

    log(`Writing upload key to ${uploadKeyPath}`);
    await AndroidCredentials.exportCertBase64(
      {
        keystorePath,
        keystorePassword: get(view, 'credentials.keystorePassword'),
        keyAlias: get(view, 'credentials.keyAlias'),
      },
      uploadKeyPath
    );
  } finally {
    try {
      await fs.unlink(keystorePath);
    } catch (err) {
      if (err.code !== 'ENOENT') {
        log.error(err);
      }
    }
  }
}
