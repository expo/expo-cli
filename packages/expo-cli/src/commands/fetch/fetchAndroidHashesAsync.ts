import * as fs from 'fs-extra';
import * as path from 'path';
import { AndroidCredentials } from 'xdl';

import { Context } from '../../credentials';
import { runCredentialsManager } from '../../credentials/route';
import { DownloadKeystore } from '../../credentials/views/AndroidKeystore';
import Log from '../../log';
import { assertSlug, Options } from './utils';

export async function actionAsync(projectRoot: string, options: Options): Promise<void> {
  const ctx = new Context();
  await ctx.init(projectRoot, {
    nonInteractive: options.parent?.nonInteractive,
  });
  const outputPath = path.resolve(projectRoot, `${ctx.manifest.slug}.tmp.jks`);
  try {
    assertSlug(ctx.manifest.slug);
    const experienceName = `@${ctx.projectOwner}/${ctx.manifest.slug}`;
    const view = new DownloadKeystore(experienceName, {
      outputPath,
      quiet: true,
    });
    await runCredentialsManager(ctx, view);
    const keystore = await ctx.android.fetchKeystore(experienceName);

    if (keystore) {
      await AndroidCredentials.logKeystoreHashes({
        keystorePath: outputPath,
        keystorePassword: keystore.keystorePassword,
        keyAlias: keystore.keyAlias,
        keyPassword: keystore.keyPassword,
      });
      Log.log(
        `\nNote: if you are using Google Play signing, this app will be signed with a different key after publishing to the store, and you'll need to use the hashes displayed in the Google Play console.`
      );
    } else {
      Log.warn('There is no valid Keystore defined for this app');
    }
  } finally {
    await fs.remove(outputPath);
  }
}
