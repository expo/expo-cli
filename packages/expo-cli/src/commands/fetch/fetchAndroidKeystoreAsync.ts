import * as path from 'path';

import { Context } from '../../credentials';
import { runCredentialsManager } from '../../credentials/route';
import { DownloadKeystore } from '../../credentials/views/AndroidKeystore';
import { assertSlug, maybeRenameExistingFileAsync, Options } from './utils';

export async function actionAsync(projectRoot: string, options: Options): Promise<void> {
  const ctx = new Context();
  await ctx.init(projectRoot, {
    nonInteractive: options.parent?.nonInteractive,
  });

  const keystoreFilename = `${ctx.manifest.slug}.jks`;
  await maybeRenameExistingFileAsync(projectRoot, keystoreFilename);
  const backupKeystoreOutputPath = path.resolve(projectRoot, keystoreFilename);
  const experienceName = `@${ctx.projectOwner}/${ctx.manifest.slug}`;

  assertSlug(ctx.manifest.slug);
  await runCredentialsManager(
    ctx,
    new DownloadKeystore(experienceName, {
      outputPath: backupKeystoreOutputPath,
      displayCredentials: true,
    })
  );
}
