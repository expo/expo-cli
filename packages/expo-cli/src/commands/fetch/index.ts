import { Command } from 'commander';

import {
  fetchAndroidHashesAsync,
  fetchAndroidKeystoreAsync,
  fetchAndroidUploadCertAsync,
} from './android';
import fetchIosCerts from './ios';

export default function(program: Command) {
  program
    .command('fetch:ios:certs [project-dir]')
    .description(
      `Fetch this project's iOS certificates/keys and provisioning profile. Writes files to the PROJECT_DIR and prints passwords to stdout.`
    )
    .asyncActionProjectDir(fetchIosCerts, true);

  program
    .command('fetch:android:keystore [project-dir]')
    .description(
      "Fetch this project's Android keystore. Writes keystore to PROJECT_DIR/PROJECT_NAME.jks and prints passwords to stdout."
    )
    .asyncActionProjectDir(fetchAndroidKeystoreAsync, true);

  program
    .command('fetch:android:hashes [project-dir]')
    .description(
      "Fetch this project's Android key hashes needed to set up Google/Facebook authentication. Note: if you are using Google Play signing, this app will be signed with a different key after publishing to the store, and you'll need to use the hashes displayed in the Google Play console."
    )
    .asyncActionProjectDir(fetchAndroidHashesAsync, true);

  program
    .command('fetch:android:upload-cert [project-dir]')
    .description(
      "Fetch this project's upload certificate needed after opting in to app signing by Google Play or after resetting a previous upload certificate."
    )
    .asyncActionProjectDir(fetchAndroidUploadCertAsync, true);
}
