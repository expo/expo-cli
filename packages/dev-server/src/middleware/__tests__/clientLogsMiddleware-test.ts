import http from 'http';

import bodyParser from 'body-parser';
import bunyan from '@expo/bunyan';
import connect from 'connect';
import fetch from 'node-fetch';

import clientLogsMiddleware from '../clientLogsMiddleware';

const headers = {
  'content-type': 'application/json',
  'device-id': '11111111-CAFE-0000-0000-111111111111',
  'session-id': '22222222-C0DE-0000-0000-222222222222',
  'device-name': 'iPhone',
};

it('logs messages from the device', async () => {
  const { server, url, logStream } = await createServerAsync();
  try {
    const response = await fetch(`${url}/logs`, {
      method: 'POST',
      headers,
      body: JSON.stringify([
        {
          count: 1,
          level: 'info',
          body: ['Hello world!'],
          includesStack: false,
          groupDepth: 0,
        },
        {
          count: 2,
          level: 'error',
          body: [
            {
              message: 'Something went wrong...',
              stack: 'App.js:3:12 in <global>',
            },
          ],
          includesStack: true,
          groupDepth: 0,
        },
        {
          // We want this to be filtered out.
          count: 3,
          level: 'info',
          body: [
            'BugReporting extraData:',
            'Object {\n' +
              '  "AppRegistry.runApplication1": "Running \\"main\\" with yada yada yada",\n' +
              '}',
          ],
          includesStack: false,
          groupDepth: 0,
        },
      ]),
    });
    expect(response.ok).toBe(true);
    expect(logStream.output).toMatchSnapshot();
  } finally {
    server.close();
  }
});

class TestLogStream {
  output: string[] = [];

  write(record: any) {
    const message = record.includesStack ? JSON.parse(record.msg).message : record.msg;
    const deviceName = record.deviceName ?? '';
    if (record.level < bunyan.INFO) {
      this.output.push(`${deviceName}: [debug] ${message}`);
    } else if (record.level < bunyan.WARN) {
      this.output.push(`${deviceName}: [info] ${message}`);
    } else if (record.level < bunyan.ERROR) {
      this.output.push(`${deviceName}: [warn] ${message}`);
    } else {
      this.output.push(`${deviceName}: [error] ${message}`);
    }
  }
}

async function createServerAsync() {
  const logStream = new TestLogStream();
  const logger = bunyan.createLogger({
    name: 'expo-test',
    streams: [
      {
        type: 'raw',
        stream: logStream,
        level: 'info',
      },
    ],
  });
  const app = connect()
    .use(bodyParser.json())
    .use(clientLogsMiddleware(logger));

  const server = http.createServer(app);
  await new Promise((resolve, reject) =>
    server.listen((error: any) => {
      if (error) reject(error);
      else resolve();
    })
  );

  const address = server.address();
  if (!address || typeof address === 'string') throw new Error('server has no port');

  return {
    server,
    url: `http://localhost:${address.port}`,
    logStream,
  };
}
