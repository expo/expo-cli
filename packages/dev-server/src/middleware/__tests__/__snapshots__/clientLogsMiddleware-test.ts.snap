// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs messages from the device 1`] = `
Array [
  "iPhone: [info] Hello world!",
  "iPhone: [error] Something went wrong...",
]
`;
